File: webapp/tauri/consumer.ts
import { invoke } from "@tauri-apps/api";
import { ConsumerConfiguration, ConsumerState, KafkaRecord } from "../models/kafka";
import { addNotification } from "../providers";
import { format, TauriError } from "./error";

export const getConsumerState = (clusterId: string, topic: string): Promise<ConsumerState> =>
  invoke<ConsumerState>("get_consumer_state", { clusterId, topic }).catch((err: TauriError) => {
    addNotification({ type: "error", title: "Get Kafka consumer state", description: format(err) });
    throw err;
  });

export const stopConsumer = (clusterId: string, topic: string): Promise<void> =>
  invoke<void>("stop_consumer", { clusterId, topic }).catch((err: TauriError) =>
    addNotification({ type: "error", title: "Stop Kafka record", description: format(err) })
  );

export const startConsumer = (clusterId: string, topic: string, config: ConsumerConfiguration): Promise<void> =>
  invoke<void>("start_consumer", { clusterId, topic, config }).catch((err: TauriError) =>
    addNotification({ type: "error", title: "Start Kafka record", description: format(err) })
  );

type GetRecordsPageResponse = {
  records: KafkaRecord[];
  nextPage?: number;
  prevPage?: number;
};

export const getRecordsPage = (
  clusterId: string,
  topic: string,
  pageNumber: number,
  query?: string
): Promise<GetRecordsPageResponse> =>
  invoke<GetRecordsPageResponse>("get_records_page", { clusterId, topic, query, pageNumber }).catch(
    (err: TauriError) => {
      addNotification({ type: "error", title: "Get Kafka records page", description: format(err) });
      throw err;
    }
  );

type ExportOptions = {
  query: string;
  outputPath: string;
  limit?: number;
  overwrite: boolean;
  parseTimestamp: boolean;
};

export const exportRecords = (clusterId: string, topic: string, options: ExportOptions): Promise<void> =>
  invoke<void>("export_records", {
    clusterId,
    topic,
    options,
  }).catch((err: TauriError) =>
    addNotification({ type: "error", title: "Export records to csv file.", description: format(err) })
  );

File: backend/src/api/consumer.rs
use log::debug;

use crate::lib::{
    consumer::{ types::ConsumerState, Consumer, ConsumerConfiguration },
    record_store::types::ExportOptions,
};

use super::{ error::Result, types::GetPageResponse, AppState };

#[tauri::command]
pub async fn start_consumer(
    cluster_id: &str,
    topic: &str,
    config: ConsumerConfiguration,
    state: tauri::State<'_, AppState>
) -> Result<()> {
    let consumer = state.get_cluster(cluster_id).await?.get_consumer(topic).await;
    Ok(consumer.start(&config).await?)
}

#[tauri::command]
pub async fn get_consumer_state(
    cluster_id: &str,
    topic: &str,
    state: tauri::State<'_, AppState>
) -> Result<ConsumerState> {
    let consumer = state.get_cluster(cluster_id).await?.get_consumer(topic).await;
    Ok(consumer.get_consumer_state().await?)
}

#[tauri::command]
pub async fn stop_consumer(cluster_id: &str, topic: &str, state: tauri::State<'_, AppState>) -> Result<()> {
    let consumer = state.get_cluster(cluster_id).await?.get_consumer(topic).await;
    Ok(consumer.stop().await?)
}

#[tauri::command]
pub async fn get_records_page(
    cluster_id: &str,
    topic: &str,
    page_number: usize,
    query: Option<&str>,
    state: tauri::State<'_, AppState>
) -> Result<GetPageResponse> {
    debug!("Get records page");
    const PAGE_SIZE: usize = 20;
    let cluster = state.get_cluster(cluster_id).await?;
    let topic_store = cluster.get_topic_store(topic).await;
    let records = topic_store.get_records(query, (page_number * PAGE_SIZE) as i64, PAGE_SIZE as i64, None)?;
    let records_count = records.len();
    Ok(GetPageResponse {
        records,
        next_page: if records_count == PAGE_SIZE {
            Some(page_number + 1)
        } else {
            None
        },
        prev_page: if page_number >= 1 {
            Some(page_number - 1)
        } else {
            None
        },
    })
}

#[tauri::command]
pub async fn export_records(
    cluster_id: &str,
    topic: &str,
    options: ExportOptions,
    state: tauri::State<'_, AppState>
) -> Result<()> {
    let store = state.get_cluster(cluster_id).await?.get_topic_store(topic).await;
    Ok(store.export_records(&options)?)
}